documents/config.md
# Config

## Link: [here](https://github.com/NordicSemiconductor/asset-tracker-cloud-docs/blob/saga/docs/cloud-protocol/Config.ts)

## Data

| Property | LwM2M     |
| -------- | --------- |
| act      | 50009.0.0 |
| actwt    | 50009.0.2 |
| mvres    | 50009.0.3 |
| mvt      | 50009.0.4 |
| loct     | 50009.0.1 |
| accath   | 50009.0.5 |
| accith   | 50009.0.8 |
| accito   | 50009.0.9 |
| nod      | **????**  |

## Description

| Property | Description                                                                                                                                                                              | Type             | Minimum | Maximum    | Examples                               | Required |
| -------- | ---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- | ---------------- | ------- | ---------- | -------------------------------------- | -------- |
| act      | Whether to enable the active mode.                                                                                                                                                       | boolean          |         |            | false                                  | Yes      |
| actwt    | In active mode: Wait this amount of seconds until sending the next update. The actual interval will be this time plus the time it takes to get a GNSS fix.                               | integer          | 1       | 2147483647 | 60                                     | Yes      |
| mvres    | Movement resolution (in seconds): After detecting movement in passive mode send an update and wait this amount of time until movement again can trigger the next update.                 | integer          | 1       | 2147483647 | 300                                    | Yes      |
| mvt      | Movement timeout (in seconds): Send update at least this often in passive mode.                                                                                                          | integer          | 1       | 2147483647 | 3600                                   | Yes      |
| loct     | Location search timeout (in seconds): Timeout for location search (GNSS fix, cellular, and WiFi positioning).                                                                            | integer          | 1       | 2147483647 | 60                                     | Yes      |
| accath   | Accelerometer activity threshold (in m/s²): Minimal absolute value for an accelerometer reading to be considered movement.                                                               | number           | 0       | 78.4532    | 10.5                                   | Yes      |
| accith   | Accelerometer inactivity threshold (in m/s²): Maximum absolute value for an accelerometer reading to be considered stillness. Must be smaller than the accelerometer activity threshold. | number           | 0       | 78.4532    | 5.2                                    | Yes      |
| accito   | Accelerometer inactivity timeout (in s): Hysteresis timeout for stillness detection. Must be smaller than the movement resolution.                                                       | number           | 0.08    | 5242.88    | 1.7                                    | Yes      |
| nod      | List of modules which should be disabled when sampling data.                                                                                                                             | array of strings |         |            | ["gnss"], ["ncell"], ["gnss", "ncell"] | Yes      |


documents/environment.md
# Environment

## Link: [here](https://github.com/NordicSemiconductor/asset-tracker-cloud-docs/blob/saga/docs/cloud-protocol/Reported.ts)

## Data

| Field | LwM2M                                                                                                                                                                      |
| ----- | -------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| temp  | 3303.0.5700                                                                                                                                                                |
| hum   | 3304.0.5700                                                                                                                                                                |
| atmp  | 3323.0.5700                                                                                                                                                                |
| ts    | 3303.0.5518 or 3304.0.5518 or 3323.0.5518 or [Timestamp Hierarchy](https://github.com/MLopezJ/asset-tracker-cloud-coiote-azure-converter-js/tree/saga#timestamp-hierarchy) |

## Details

| Field | Description                                              | Type    | Minimum       | Maximum | Required |
| ----- | -------------------------------------------------------- | ------- | ------------- | ------- | -------- |
| temp  | Temperature reading from external sensor                 | number  |               |         | Yes      |
| hum   | Humidity reading from external sensor                    | number  | 1             | 100     | Yes      |
| atmp  | Atmospheric pressure reading from external sensor in kPa | number  | 0             |         | Yes      |
| ts    | Timestamp as Unix epoch with millisecond precision (UTC) | integer | 1234567890123 |         | Yes      |


src/schemas/Config.ts
import { Static, Type } from '@sinclair/typebox'
import { $id } from './$id.js'

/**
 * @see https://github.com/NordicSemiconductor/asset-tracker-cloud-docs/blob/saga/docs/cloud-protocol/Config.ts
 */
export const Config = Type.Object(
	{
		act: Type.Boolean({
			description: 'Whether to enable the active mode.',
			examples: [false],
		}),
		actwt: Type.Integer({
			description:
				'In active mode: Wait this amount of seconds until sending the next update. The actual interval will be this time plus the time it takes to get a GNSS fix.',
			minimum: 1,
			maximum: 2147483647,
			examples: [60],
		}),
		mvres: Type.Integer({
			description:
				'Movement resolution (in seconds): After detecting movement in passive mode send an update and wait this amount of time until movement again can trigger the next update.',
			minimum: 1,
			maximum: 2147483647,
			examples: [300],
		}),
		mvt: Type.Integer({
			description:
				'Movement timeout (in seconds): Send update at least this often in passive mode.',
			minimum: 1,
			maximum: 2147483647,
			examples: [3600],
		}),
		loct: Type.Integer({
			description:
				'Location search timeout (in seconds): Timeout for location search (GNSS fix, cellular, and WiFi positioning).',
			minimum: 1,
			maximum: 2147483647,
			examples: [60],
		}),
		accath: Type.Number({
			description:
				'Accelerometer activity threshold (in m/s²): Minimal absolute value for an accelerometer reading to be considered movement.',
			minimum: 0,
			maximum: 78.4532,
			examples: [10.5],
		}),
		accith: Type.Number({
			description:
				'Accelerometer inactivity threshold (in m/s²): Maximum absolute value for an accelerometer reading to be considered stillness. Must be smaller than the accelerometer activity threshold.',
			minimum: 0,
			maximum: 78.4532,
			examples: [5.2],
		}),
		accito: Type.Number({
			description:
				'Accelerometer inactivity timeout (in s): Hysteresis timeout for stillness detection. Must be smaller than the movement resolution.',
			minimum: 0.08,
			maximum: 5242.88,
			examples: [1.7],
		}),
		nod: Type.Array(Type.String({ minLength: 1 }), {
			description:
				'List of modules which should be disabled when sampling data.',
			type: 'array',
			minItems: 0,
			examples: [['gnss'], ['ncell'], ['gnss', 'ncell']],
		}),
	},
	{
		$schema: 'http://json-schema.org/draft-07/schema#',
		$id: $id('config'),
		description: 'Configures the device',
	},
)

export type config = Static<typeof Config>
